util/image_viewer/slip_viewer/tile_generation.py

```diff
@@ -114,6 +114,14 @@ def get_flex_image_multipanel(
             pass
     beam_center /= npanels / 1e-3
 
+    # Calculate 2D origin, fast and slow vectors for detector projected onto
+    # a best fit frame. FIXME this should be done once for the detector, not
+    # here for every image we want to display!
+    try:
```

Comment:
EAFP is OK, but wonder if this is a little harder to read than just a "hasattr" or even setting the attribute to `None` when the object is created (which would involve a little more exploration but helps to leave breadcrumbs for a longer term improvement)

Reply:
oh the `FIXME` comment is out of date too btw. It came from the first commit where `project_2d` was called every time this function run rather than checking the detector for pre-calculated results. I did do this in a rush yesterday...

Reply:
`panels` is a misnomer here, AFAICT. It is actually a `Detector` object. We _could_ assign this attribute when the experiments are first loaded in, to ensure it is there whenever it matters.

An alternative is what I suggested at the bottom of https://github.com/cctbx/dxtbx/pull/224#issue-494737806, which is that the `project_2d` function gets moved to the `Detector` object itself. Then any detector can tell a viewer how to lay out its panels in 2D. This would have the advantage of providing a method that could be overridden in special cases, like that of the P12M, avoiding the special case code here in the image viewer.

Reply:
hmm, but actually it would probably have to be in the `Format`, because I don't think a `Detector` knows what detector it is...

Reply:
until then: 
`projected_axes = getattr(panels, "projected_2d", None)`
replaces the 4 lines